contracts/PeeranhaCommunityToken.sol

```diff
@@ -0,0 +1,138 @@
+//SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import "./interfaces/IPeeranhaCommunityToken.sol";
+import "./interfaces/IPeeranhaCommunityTokenFactory.sol";
+import "./libraries/CommonLib.sol";
+import "./base/NativeMetaTransaction.sol";
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/extensions/IERC20MetadataUpgradeable.sol";
+
+// transfer: 
+//    payable(userAddress).transfer(amount);
+//    IERC20Upgradeable(tokenAddress).transfer(userAddress, 2);
+//
+// balace:
+//    userAddress.balance
+//    IERC20Upgradeable(tokenAddress).balanceOf(userAddress)
+
+
+contract PeeranhaCommunityToken is IPeeranhaCommunityToken, NativeMetaTransaction {  
+
+  // sumAccruedTokens -> free tokens (added - pool)
+  // sumSpentTokens -> active tokens tokens (pools - give reward)
+  // frezeTokens - not taked pool
+  struct CommunityToken {
+    string name;
+    string symbol;
+    address tokenAddress;
+    uint256 maxRewardPerPeriod;
+    uint256 activeUsersInPeriod;
+    uint256 maxRewardPerUser;
+    uint256 frezeTokens;
+    uint256 createTime;
+    address peeranhaCommunityTokenFactoryAddress;
+  }
+
+  struct CommunityTokenContainer {
+    CommunityToken info;
+    mapping(uint16 => uint256) periodPool;
+  }
+
+  CommunityTokenContainer communityTokenContainer;
+
+  event AddBalance(uint256 indexed amount);   // name
+
+  constructor(address tokenAddress, uint256 maxRewardPerPeriod, uint256 activeUsersInPeriod, address peeranhaCommunityTokenFactoryAddress) {
+    if (tokenAddress == address(0))
+      communityTokenContainer.info.tokenAddress = address(0x0000000000000000000000000000000000001010); // chack Address !
+    else
+      communityTokenContainer.info.tokenAddress = tokenAddress;
+    communityTokenContainer.info.name = IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).name();
+    communityTokenContainer.info.symbol = IERC20MetadataUpgradeable(communityTokenContainer.info.tokenAddress).symbol();
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.activeUsersInPeriod = activeUsersInPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerPeriod / activeUsersInPeriod;
+    communityTokenContainer.info.createTime = CommonLib.getTimestamp();
+    communityTokenContainer.info.peeranhaCommunityTokenFactoryAddress = peeranhaCommunityTokenFactoryAddress;
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override
+      view
+      returns (address sender)
+  {
+    return NativeMetaTransaction._msgSender();
+  }
+
+  function addBalance(uint256 amount) external {  // name
+    address owner = _msgSender();
+    IERC20Upgradeable(communityTokenContainer.info.tokenAddress).transferFrom(owner, address(this), amount);
+
+    emit AddBalance(amount);
```

Comment:
Do we need this event for indexer to work properly?
It will be much more convenient for community admins just to transfer tokens to address of community reward contract from their wallet.